src/AST.h

```diff
@@ -1,32 +1,39 @@
 #pragma once
+#include<string>
+
+#include <string>
 
 #include <string>
 
 class Location {
+
 public:
-	Location(int firstLine, int firstCol, int lastLine, int lastCol)
-		: first_line(firstLine), first_col(firstCol), last_line(lastLine), last_col(lastCol) {
+	Location(int firstLine, int firstCol, int lastLine, int lastCol,
+			const std::string &filename)
+		: first_line(firstLine), first_col(firstCol), last_line(lastLine),
+		last_col(lastCol), filename(filename) {
 	}
 
+	std::string fileName() const { return filename; }
 	int firstLine() const { return first_line; }
 	int firstColumn() const { return first_col; }
 	int lastLine() const { return last_line; }
 	int lastColumn() const { return last_col; }
 
 
 	static const Location NONE;
-;
 private:
 	int first_line;
 	int first_col;
 	int last_line;
 	int last_col;
+	std::string filename;
```

Comment:
It feels a bit intense to store the filename for each AST node. What  you're really using this for is to determine if two assignments belong to the same file.
I've been contemplating ways of managing assignments in a better way, which could help towards this. My idea thus far is to parse included files into separate `FileModule` instances, and resolve the actual assignments during evaluation. This would also make OpenSCAD's special assignment rules more explicit and isolated in our source code. See #1743 

The above PR may take time to push through though, so we may want to implement your solution in the interim.

Reply:
I could have used a flag called "included" to get the same result for just now.

I picked the filename, as it seems more future prove in regards to general error handling.
(e.g. in which include did the assertion fail?)

Also, that specific line is from #2044, where the argument was about failing tests, not the general idea.

I see two ways:
* we can close this pull request and I create a new one based  on this with a "included flag"
or
*  I change the std::string to an int, storing the association from int to string in a look up table

I have personal attachment to the string solution specifically, if you have bigger plans for the long run, I do not want to get in your way. Well, the more I talk: the included flag sounds sensible.

Reply:
Keep in mind that we may have to handle reassignment in a smart way. Haven't thought about that yet..
Nevertheless, I think the string solution is still the better one as it's really explicit and should be easy to refactor later.

Reply:
Could we handle that via a shared pointer created in the parser? I guess having the filename around would also help with error messages.

Reply:
shared pointer sounds perfect. 32/64 bits more per AST node is cheap for the benefits in error reporting we get in knowing the file name.

But why specially created in the parser?
For e.g. FileModule could (and should) normalize the paths and could then provide use with a pointer.

I will look into it.

Reply:
@MichaelPFrey Going down that route, would a pointer to the owning `FileModule` be even better, to make the connection explicit? Would probably have to be a `weak_ptr` or smth. to avoid circular references.